src/main/java/de/tum/cit/aet/artemis/programming/service/localvc/LocalVCServletService.java

```diff
@@ -715,6 +759,125 @@ public static String getDefaultBranchOfRepository(Repository repository) {
+        }
+    }
+
+    /**
+     * Updates the VCS access log for clone and pull actions performed over SSH.
+     * <p>
+     * This method logs access information based on the SSH session and the root directory of the repository.
+     * It determines the repository action (clone or pull) based on the number of offers (`cntOffered`) and
+     * fetches participation details from the local VC repository URI.
+     *
+     * @param session    the {@link ServerSession} representing the SSH session.
+     * @param rootDir    the {@link Path} to the root directory of the repository.
+     * @param cntOffered the number of objects offered by the client in the operation, used to determine
+     *                       if the action is a clone (if 0) or a pull (if greater than 0).
+     */
+    @Async
+    public void updateVCSAccessLogForCloneAndPullSSH(ServerSession session, Path rootDir, int cntOffered) {
+        try {
+            if (session.getAttribute(SshConstants.USER_KEY).getName().equals("buildjob_user")) {
+                return;
+            }
+            RepositoryActionType repositoryActionType = getRepositoryActionReadType(cntOffered);
+            var parti = getExerciseParticipationFromLocalVCRepositoryUri(getLocalVCRepositoryUri(rootDir));
+            vcsAccessLogService.ifPresent(service -> service.updateRepositoryActionType(parti, repositoryActionType));
+        }
+        catch (Exception ignored) {
+        }
+    }
+
+    /**
+     * Adds a failed VCS access attempt to the log.
+     * <p>
+     * This method logs a failed clone attempt, associating it with the user and participation retrieved
+     * from the incoming HTTP request. It assumes that the failed attempt used password authentication.
+     *
+     * @param servletRequest the {@link HttpServletRequest} containing the HTTP request data.
+     */
+    public void addVCSAccessLogFailedAttempt(HttpServletRequest servletRequest) {
+        try {
+            User user = getUserFromRequest(servletRequest);
+            var participation = getExerciseParticipationFromRequest(servletRequest);
+            vcsAccessLogService.ifPresent(service -> service.storeAccessLog(user, participation, RepositoryActionType.CLONE_FAIL, AuthenticationMechanism.PASSWORD, "", ""));
+        }
+        catch (LocalVCAuthException ignored) {
+        }
+    }
+
+    /**
+     * Determines the repository action type for read operations (clone or pull).
+     * <p>
+     * This method returns a {@link RepositoryActionType} based on the number of objects offered.
+     * If no objects are offered (0), it is considered a clone; otherwise, it is a pull action.
+     *
+     * @param cntOffered the number of objects offered to the client in the operation.
+     * @return the {@link RepositoryActionType} based on the number of objects offered (clone if 0, pull if greater than 0).
+     */
+    public RepositoryActionType getRepositoryActionReadType(int cntOffered) {
```

Comment:
I think we normally do not abbreviate parameter names

Reply:
yea, I only carried through the name from the overridden method were it comes from. I renamed it to `clientOffered`